### PR TITLE
[release/5.0] Backport System.Speech to compat pack

### PR DESCRIPTION
**Pull request by danmoseley**
Backport System.Speech including its entry in the windows compat pack.
Original PR https://github.com/dotnet/runtime/pull/45941

## Customer Impact

The System.Speech assembly on .NET Framework does not work on .NET Core/.NET 5 because it uses private reflection. Many customers rely on generating speech from .NET code to make their apps or scripts accessible to them or their users. They have given us clear and extensive feedback on the impact this has on them, eg in the issue https://github.com/dotnet/wpf/issues/2935. In particular this means they cannot generate speech from Powershell scripts, as recent versions of Powershell depend on .NET Core.

This change backports the port of System.Speech that we have completed in master, so that we can ship it in an updated Windows Compatibility Pack and enable this speech functionality for our customers using .NET 2.1+.

## Testing

* Unit tests that give ~50% coverage. That isn't great, but they cover the key synthesis/recognition paths. Much of the uncovered code are paths like event handlers, or code relating to generating or parsing custom grammars, etc that is not as relevant to generating basic speech which is the key use case we are trying to enable. There aren't quick wins to increase the coverage substantially, and the original tests are not feasible to port due to their structure.
* Manually tested the System.Speech package as follows:  net472->ns20, net472->net472, netcoreapp2.1->ns20, netcoreapp2.2->ns20, net50->ns2.0. Where a P2P is shown, code was exercised in both. Also verified a build error for netcoreapp2.0.
* Manually tested the Windows Compat pack package (containing it) as follows: net50->ns20, netcoreapp21->ns20

## Risk

The code forms its own library, not previously shipped, so it's highly unlikely that existing .NET Core users would be broken by shipping this library. If there was a break, it would presumably relate to a packaging issue (since it's included into the compatibility pack). We tested key scenarios around consuming the package.

### CODE REVIEW

File: src/libraries/pkg/baseline/packageIndex.json
Comment (danmosemsft) [inline src/libraries/pkg/baseline/packageIndex.json:6253]: Is it correct that this is 5.0.2? If I use 5.0.0 I get an error: `PackageIndex from C:\git\runtime\src\libraries\pkg\baseline\packageIndex.json indicates that assembly version 4.0.0.0 is contained in non-stable package version 5.0.0 which differs from this package version 5.0.2`

I'm not sure where 5.0.2 comes from, as eg., System.Diagnostics.EventLog has 5.0.0.

**Comment by danmosemsft** [comment]
Test failure are https://github.com/dotnet/runtime/issues/43981 and https://github.com/dotnet/runtime/issues/11063

File: src/libraries/pkg/baseline/packageIndex.json
Comment (ericstj) [reply src/libraries/pkg/baseline/packageIndex.json:6253]: You should list the first package version that this assembly version appeared in.  This should be 5.0.0

Comment (danmosemsft) [reply src/libraries/pkg/baseline/packageIndex.json:6253]: That was my understanding but if I try
```
    "System.Speech": {
      "BaselineVersion": "5.0.0",
      "InboxOn": {
        "net45": "4.0.0.0"
      },
      "AssemblyVersionInPackageVersion": {
        "4.0.0.0": "5.0.0"
      }
    },
```
then I get 
```
Packaging.targets(1112,5): error : PackageIndex from C:\git\runtime\src\libraries\pkg\baseline\packageIndex.json indicates that assembly version 4.0.0.0 is contained in non-stable package version 5.0.0 which differs from this package version 5.0.2. [C:\git\runtime\src\libraries\System.Speech\pkg\System.Speech.pkgproj]
```
I'm not sure where 5.0.2 is coming from.

Comment (ericstj) [reply src/libraries/pkg/baseline/packageIndex.json:6253]: Please set PackageVersion to 5.0.0 in Directory.Build.props for this library.  Normally we set PackageVersion in servicing for libraries, otherwise they'll default to the current ProductVersion which increments in servicing.  This package is special since you're shipping it for the first time in servicing.

**Comment by ericstj** [comment]
I suspect you'll need to cherry-pick https://github.com/dotnet/runtime/pull/46598/commits/bc24a14b3b622405302bcc4b6cdb777127b42e30 from https://github.com/dotnet/runtime/pull/46688 to get package testing to pass.

**Comment by danmosemsft** [comment]
@ericstj apparently I didn't, unless it's something not run in CI.

**Comment by danmosemsft** [comment]
Running pkg/tests locally i do get a bunch of errors like
```
C:\git\50runtime\artifacts\bin\testPkg\projects\System.Diagnostics.DiagnosticSource\netcoreapp1.0\project.csproj : error NU3005: Package 'Microsoft.NETCore.Platforms 1.0.2' from source 'C:\Program Files\dotnet\sdk\NuGetFallbackFolder': The package signature file entry is invalid. The central directory header field 'compression method' has an invalid value (8). [C:\git\50runtime\src\libraries\pkg\test\testPackages.proj]
```

**Comment by danmosemsft** [comment]
Oh, silly mistake, I didn't remove the 5.0.2.

**Comment by ericstj** [comment]
> @ericstj apparently I didn't, unless it's something not run in CI.

Did you add the package to https://github.com/dotnet/runtime/blob/988fa3b75f8d0d1f11ee0c2856ecdfeefea2cd27/src/libraries/libraries-packages.proj#L21?  If not it won't build during servicing.  

Double check the steps here: https://github.com/dotnet/runtime/blob/release/5.0/docs/project/library-servicing.md

**Comment by danmosemsft** [comment]
OK added to libraries-packages.proj. Also found it did'nt work on .NET Framework -- fixed public key to MicrosoftShared to fix that.

Strangely something is updating packageIndex.json by itself to add lots of "5.0.2" entries. I'll see whether it's the build..

**Comment by danmosemsft** [comment]
Yes, when I do `build clr+libs+libs.tests -rc release -allconfigurations -pack` I get this edit created for me:
```diff
diff --git a/src/libraries/pkg/baseline/packageIndex.json b/src/libraries/pkg/baseline/packageIndex.json
index cff6d7f372d..176476355a1 100644
--- a/src/libraries/pkg/baseline/packageIndex.json
+++ b/src/libraries/pkg/baseline/packageIndex.json
@@ -23,7 +23,8 @@
     "Microsoft.Bcl.AsyncInterfaces": {
       "StableVersions": [
         "1.0.0",
-        "5.0.0"
+        "5.0.0",
+        "5.0.2"
       ],
       "BaselineVersion": "5.0.0",
       "InboxOn": {},
@@ -137,7 +138,8 @@
         "2.0.0",
         "2.0.1",
         "2.1.0",
-        "5.0.0"
+        "5.0.0",
+        "5.0.2"
       ],
       "BaselineVersion": "5.0.0",
       "InboxOn": {},
@@ -170,7 +172,8 @@
         "3.1.0",
         "3.1.1",
         "3.1.2",
-        "5.0.0"
+        "5.0.0",
+        "5.0.2"
       ],
       "InboxOn": {},
       "AssemblyVersionInPackageVersion": {
@@ -199,7 +202,8 @@
         "3.1.0",
         "3.1.1",
         "3.1.2",
-        "5.0.0"
+        "5.0.0",
+        "5.0.2"
       ],
       "InboxOn": {},
etc etc
```
is that an issue?

**Comment by janvorli** [comment]
> Yes, when I do `build clr+libs+libs.tests -rc release -allconfigurations -pack` I get this edit created for me

I've also noticed it happens whenever I try to build the 5.0 branch locally.

**Comment by danmosemsft** [comment]
Need to cherry pick https://github.com/dotnet/runtime/pull/46841

**Comment by danmosemsft** [comment]
Marking no merge until I can cherry pick #46841 shortly

File: src/libraries/System.Speech/tests/SpeechRecognizerTests.cs
Comment (danmosemsft) [reply src/libraries/System.Speech/tests/SpeechRecognizerTests.cs:28]: Yes, probably - I noticed that after I merged the other PR. If it's OK, I'd like to leave it as it's harmless, to avoid having to rerun all the outer loop again. I'll clean it up in master.

**Comment by danmosemsft** [comment]
@ericstj any other feedback? if you can approve, we can presumably merge.

**Review by ericstj — APPROVED** [review]
LGTM.  I think @Anipik needs to merge this once he does branding and will need to address merge conflicts.

**Comment by danmosemsft** [comment]
OK thanks! @Anipik I will be happy to validate this when we have product installer that includes it.

**Comment by Anipik** [comment]
> OK thanks! @Anipik I will be happy to validate this when we have product installer that includes it.

i am waiting on this https://github.com/dotnet/runtime/pull/46888 to be green. After that i will start merging the changes.

**Review by Anipik — APPROVED** [review]

**Comment by Anipik** [comment]
> Yes, when I do build clr+libs+libs.tests -rc release -allconfigurations -pack I get this edit created for me:

This is being fixed here https://github.com/dotnet/runtime/pull/46224

**Pull request merged**
